src/backend/common/kernel_cache.cpp

```diff
@@ -82,15 +72,51 @@ Kernel findKernel(const string& kernelName, const vector<string>& sources,
         tInstance += ">";
     }
 
-    int device    = detail::getActiveDeviceId();
-    Kernel kernel = lookupKernel(device, tInstance, sources);
+    const bool notJIT = !sourceIsJIT;
+
+    vector<string> everything;
+    everything.reserve(1 + (notJIT * (sources.size() + options.size())));
+    everything.push_back(tInstance);
+    if (notJIT) {
+        // This code path is only used for regular kernel compilation
+        // since, jit funcName(kernelName) is unique to use it's hash
+        // for caching the relevant compiled/linked module
+        everything.insert(everything.end(), sources.begin(), sources.end());
+        everything.insert(everything.end(), options.begin(), options.end());
+    }
+
+    const string moduleKey = std::to_string(deterministicHash(everything));
+    const int device       = detail::getActiveDeviceId();
+    Module currModule      = findModule(device, moduleKey);
 
-    if (kernel.getModule() == nullptr || kernel.getKernel() == nullptr) {
-        kernel = compileKernel(kernelName, tInstance, sources, compileOpts,
-                               isKernelJIT);
-        cacheKernel(device, tInstance, kernel);
+    if (currModule.get() == nullptr) {
+#if defined(AF_CUDA) && defined(AF_CACHE_KERNELS_TO_DISK)
+        currModule = loadModuleFromDisk(device, moduleKey, sourceIsJIT);
+        if (currModule.get() == nullptr) {
+            currModule = compileModule(moduleKey, sources, options, {tInstance},
+                                       sourceIsJIT);
+        }
+#else
+        currModule = compileModule(moduleKey, sources, options, {tInstance},
+                                   sourceIsJIT);
+#endif
+        std::unique_lock<shared_timed_mutex> writeLock(getModuleMutex(device));
+        auto& cache = getCache(device);
+        auto iter   = cache.find(moduleKey);
+        if (iter == cache.end()) {
+            // If not found, this thread is the first one to compile this
+            // kernel. Keep the generated module.
+            getCache(device).emplace(moduleKey, currModule);
+        } else {
+            currModule.unload();  // dump the current threads extra compilation
+            currModule = iter->second;
+        }
     }
-    return kernel;
+#if defined(AF_CUDA)
+    return getKernel(currModule, tInstance, sourceIsJIT);
```

Comment:
we can just do a substring to find the kernel name in tInstance. That way we can send the same variable to both. Not a big deal.

Reply:
That is not possible unfortunately, it would rather increase the logic since kInstance won't have < always. it is only present for regular kernels and that too for CUDA only